lib/private/MembershipManager.php

```
@@ -0,0 +1,670 @@
+			$qb->andWhere($qb->expr()->eq($alias.'.account_id',
+				$qb->createNamedParameter($userId)));
+		} else {
+			// We need to join with accounts table, since we miss information on accountId
+			// We need to join with backend group table, since we miss information on backendGroupId
+			$qb->innerJoin($alias, $this->accountMapper->getTableName(),
+				'a', $qb->expr()->eq('a.id', $alias.'.account_id'));
+			$qb->innerJoin($alias, $this->groupMapper->getTableName(),
+				'g', $qb->expr()->eq('g.id', $alias.'.backend_group_id'));
+
+			// Apply predicate on user_id in accounts table
+			$qb->where($qb->expr()->eq('a.user_id', $qb->createNamedParameter($userId)));
+
+			// Apply predicate on group_id in backend groups table
+			$qb->andWhere($qb->expr()->eq('g.group_id', $qb->createNamedParameter($groupId)));
+		}
+
+		// Place predicate on membership_type
+		$qb->andWhere($qb->expr()->eq($alias.'.membership_type', $qb->createNamedParameter($membershipType)));
+		$resultArray = $qb->execute()->fetchAll();
+		return !empty($resultArray);
+	}
+
+
+	/**
+	 * Add user to the group with specific membership type $membershipType.
+	 *
+	 * //FIXME: Can we use INSERT INTO ... SELECT .. FROM .. ?
+	 *
+	 * @param int $accountId - internal id of an account
+	 * @param int $backendGroupId - internal id of backend group
+	 * @param string $membershipType
+	 *
+	 * @return boolean
+	 */
+	private function addGroupMember($accountId, $backendGroupId, $membershipType) {
+		$qb = $this->db->getQueryBuilder();
+
+		$qb->insert($this->getTableName())
+			->values([
+				'backend_group_id' => $qb->createNamedParameter($backendGroupId),
+				'account_id' => $qb->createNamedParameter($accountId),
+				'membership_type' => $qb->createNamedParameter($membershipType),
+			]);
+
+		try {
+			$qb->execute();
+			return true;
+		} catch (UniqueConstraintViolationException $e) {
+			// FIXME: is this correct?
+			\OC::$server->getLogger()->logException($e, ['app' => 'core']);
```

@DeepDiver1975 is this correct way of logging this exception?

---

We should check a couple of things:
* Does this class need to log something? Taking into account this is the only log here, probably not.
* Does this class need to handle that exception? I have my doubts about it.

The other option would be to not catch the exception and let it bubble up in order to handle it in other place. Note that exposing a DB exception and handle it near the output response might not be good.

I'm not saying this is bad, but there might be better options. My point is that having to log something without injecting a logger is something to be avoided, as well as injecting a class (the logger) that doesn't seem to be a clear dependency (because it's only used here). Note that swallowing the exception without doing anything is a worse option.

---

Hmm, makes sense. In fact, we dont handle also foreign constrain violations, so maybe this one we should not either.